### PR TITLE
Update find_and_lock and find_and_unlock methods

### PR DESCRIPTION
**Pull request by dks17**
Changes:
- removed `locking_name_length` parameter (looks unnecessary due to ability to define own `locking_name_generator`)
- added parameter validation for `locker` method
- updated searching criteria for `find_and_lock`, `find_and_unlock` and `unlocked` methods
- now `find_and_unlock` accepts options to find a document (a document locking_name_field could be change during the lock, which made it impossible to unlock a document)
- added index by `locking_name` to the tests

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 ### 2.0.0 (Next)
 
 * Your contribution here.
+* [#79](https://github.com/mongoid/mongoid-locker/pull/79): Update find_and_lock and find_and_unlock methods - [@dks17](https://github.com/dks17).
```
Comment (dblock) [inline CHANGELOG.md]: Nitpick: these really improve performance as far as I can see, you may want to say "Improved performance of `find_and_lock` and `find_and_unlock`" instead.

File: lib/mongoid/locker.rb [hunk 2]
```diff
@@ -287,8 +295,8 @@ def lock!(opts)
       end
     end
 
-    def unlock!
-      Mongoid::Locker::Wrapper.find_and_unlock(self)
+    def unlock!(opts)
```
Comment (dblock) [inline lib/mongoid/locker.rb]: Generally default opts, `opts = {}`, especially if this is an external interface change.

File: spec/support/configurations_shared_context.rb [hunk 3]
```diff
@@ -10,7 +10,11 @@ class User
       field :locked_at, type: Time
 
       field :age, type: Integer
+
+      index({ _id: 1, locking_name: 1 }, name: 'mongoid_locker_index', sparse: true, unique: true, expire_after_seconds: lock_timeout)
     end
+
+    User.create_indexes
```
Comment (dblock) [inline spec/support/configurations_shared_context.rb]: Depending on the `database_cleaner` strategy the collections may be dropped (I didn't check) between tests, not just across `:context`. I think we want to move the `User` class out in its own file loaded once, and the creation of indexes along with `reset!` into a `before do`.

**Comment by dks17** [comment]
Thank you.

File: CHANGELOG.md (hunk 1 above)
Comment (dks17) [reply CHANGELOG.md]: Accepted.

File: lib/mongoid/locker.rb (hunk 2 above)
Comment (dks17) [reply lib/mongoid/locker.rb]: `unlock!` is protected method, and it passes `opts` into `find_and_unlock` where `opts` parameter is mandatory.

File: spec/support/configurations_shared_context.rb (hunk 3 above)
Comment (dks17) [reply spec/support/configurations_shared_context.rb]: We didn't use `database_cleaner` gem here, if you meant it.
This `index` was added with only one purpose - check that it works before it is added into `README`. So, the index works and does not impact on anything.
We can apply `Mongoid.purge!` if it is really needed here.

Not sure that understood your idea properly.

Comment (dblock) [reply spec/support/configurations_shared_context.rb]: Yeah ignore me. I was having a conversation with myself in my head without looking at the whole code ;) All is good.

**Pull request merged**